enhancements/cluster-api/cluster-autoscaler-integration-with-openshift-cluster-api.md

```diff
@@ -0,0 +1,751 @@
+  the Cluster Autoscaler.
+
+#### MachineSet Validating Admission Policy Changes
+
+To ensure that the Cluster Autoscaler can update the `replicas` field of any Cluster API
+MachineSet, there will need to be a validating admission policy. This is required because
+the Cluster Autoscaler will be configured to only interact with Cluster API resources,
+regardless of authority. Furthermore, due to the natue of authoritative resources there will
+be restrictions around how users, and applications, can update non-authoritative resources.
+In cases where a Machine API MachineSet is authoritative, updates to the Cluster API mirror
+of that resource would be rejected.
+
+A validating admission policy to allow the Cluster Autoscaler service account to have write
+access to MachineSets will be created. The following examples show what the possible
+validating admission policies might look like:
+
+```
+matchConstraints:
+  resourceRules:
+  - apiGroups:   ["machine.openshift.io"]
+    apiVersions: ["v1beta1"]
+    operations:  ["UPDATE"]
+    resources:   ["MachineSet"]
+matchConditions:
+  # Only check requests coming from the cluster autoscaler service account.
+  - name: "check-only-cluster-autoscaler-service-account-requests"
+    expression: '(request.userInfo.username in [
+            "system:serviceaccount:openshift-machine-api:cluster-autoscaler",
+            ])'
+validations:
+  - expression: 'object.spec.replicas != oldObject.spec.replicas'
+    messageExpression: "Requested replica change is the same as current value"
+```
+
+```
+matchConstraints:
+  resourceRules:
+  - apiGroups:   ["cluster.x-k8s.io"]
+    apiVersions: ["v1beta1"]
+    operations:  ["UPDATE"]
+    resources:   ["MachineSet"]
+matchConditions:
+  # Only check requests coming from the cluster autoscaler service account.
+  - name: "check-only-cluster-autoscaler-service-account-requests"
+    expression: '(request.userInfo.username in [
+            "system:serviceaccount:openshift-machine-api:cluster-autoscaler",
+            ])'
+validations:
+  - expression: 'object.spec.replicas != oldObject.spec.replicas'
+    messageExpression: "Requested replica change is the same as current value"
+```
+
+### Risks and Mitigations
+
+One of the possible risks associated with this implementation is conflicting
+MachineAutoscaler resources; where a user has created two resources
+that reference both the authoritative and non-authoritative MachineSet. This
+could create a race condition where updating the minimum and maximum size
+values will lead to an inaccurate update to both MachineSets.
+
+To address this possible risk, the Cluster Autoscaler Operator will only write to the
+authoritative MachineSet resource. A user may create a MachineAutoscaler that references
+either the authoritative or non-authoritative resource in its `targetScaleRef` field,
+but the Cluster Autoscaler Operator will only update the authoritative resource.
```

Comment:
This doesn't align with the error condition above? Are we going to take first created as the correct object? Can we include a VAP that prevents multiple objects using the same Name in the ref?

Reply:
which part doesn't align? (i thought i had caught the change)

> Are we going to take first created as the correct object?

yes, this is essentially how it will work.

> Can we include a VAP that prevents multiple objects using the same Name in the ref?

that sounds like a good upgrade, i'll add something about it.